File: src/Scripts/OGTextField.js
#pragma strict

@script AddComponentMenu ("OpenGUI/TextField")

import System.Text.RegularExpressions;

class OGTextField extends OGWidget {
	enum RegExPreset {
		None,
		OnlyNumbers,
		OnlyNumbersAndPeriod,
		OnlyASCII,
		NoSpaces
	}
	
	var locked : boolean = false;
	var text : String = "";
	var maxLength : int = 30;
	var regex : String;
	var regexPreset : RegExPreset;
	
	@HideInInspector var editor : TextEditor;
	@HideInInspector var currentPreset : RegExPreset = RegExPreset.None;
	
	override function UpdateWidget () {		
		if ( regexPreset != currentPreset ) {
			currentPreset = regexPreset;
			
			if ( currentPreset == RegExPreset.None ) {
				regex = "";
		
			} else if ( currentPreset == RegExPreset.OnlyNumbers ) {
				regex = "^0-9";
				
			} else if ( currentPreset == RegExPreset.OnlyASCII ) {
				regex = "^a-zA-Z0-9";
				
			} else if ( currentPreset == RegExPreset.NoSpaces ) {
				regex = " ";
				
			} else if ( currentPreset == RegExPreset.OnlyNumbersAndPeriod) {
				regex = "^0-9.";
				
			}
		}
	}
	
	override function Draw ( x : float, y : float ) {
		if ( !text ) { text = ""; }
		if ( !guiStyle ) { guiStyle = GUI.skin.textField; }
		
		if ( locked ) {
			GUI.TextField ( Rect ( x, y, transform.localScale.x, transform.localScale.y ), text, maxLength, guiStyle );
			
		} else {			
			text = GUI.TextField ( Rect ( x, y, transform.localScale.x, transform.localScale.y ), text, maxLength, guiStyle );
		
		}
		
		editor = GUIUtility.GetStateObject ( TextEditor, GUIUtility.keyboardControl ) as TextEditor;
		
		if ( regex != "" && regex != "\\" && regexPreset != RegExPreset.None ) {
			text = Regex.Replace ( text, "[" + regex + "]", "" );
		}
	}
}

File: C#/Scripts/OGTextField.cs
using UnityEngine;
using System.Collections;
using System.Text.RegularExpressions;

[AddComponentMenu ("OpenGUI/TextField")]
public class OGTextField : OGWidget {

	public enum RegExPreset {
		None,
		OnlyNumbers,
		OnlyNumbersAndPeriod,
		OnlyASCII,
		NoSpaces
	}
	
	public string text = "";
	public int maxLength = 30;
	public string regex;
	public RegExPreset regexPreset;
	
	TextEditor editor;
	RegExPreset currentPreset = RegExPreset.None;
	
	public override void UpdateWidget (){		
		if ( regexPreset != currentPreset ) {
			currentPreset = regexPreset;
			
			if ( currentPreset == RegExPreset.None ) {
				regex = "";
		
			} else if ( currentPreset == RegExPreset.OnlyNumbers ) {
				regex = "^0-9";
				
			} else if ( currentPreset == RegExPreset.OnlyASCII ) {
				regex = "^a-zA-Z0-9";
				
			} else if ( currentPreset == RegExPreset.NoSpaces ) {
				regex = " ";
				
			} else if ( currentPreset == RegExPreset.OnlyNumbersAndPeriod) {
				regex = "^0-9.";
				
			}
		}
	}
	
	public override void Draw (  float x ,   float y   ){
		if ( guiStyle== null ) { guiStyle = GUI.skin.textField; }
					
		text = GUI.TextField ( new Rect( x, y, transform.localScale.x, transform.localScale.y ), text, maxLength, guiStyle );
		editor = GUIUtility.GetStateObject ( typeof(TextEditor), GUIUtility.keyboardControl ) as TextEditor;
		
		if ( regex != "" && regex != "\\" ) {
			text = Regex.Replace ( text, "[" + regex + "]", "" );
		}
	}
}
